### PR TITLE
🌐 restore corresponding source branch firstly

### PR DESCRIPTION
**Pull request by OsmondJiang**
build loc xxx branch
  - restore source xxx branch
  - if source xxx branch doesn't exist, restore `master` branch

### CODE REVIEW

**Comment by OsmondJiang** [comment]
#4357

File: src/docfx/localization/LocalizationUtility.cs [hunk 1]
```diff
@@ -131,11 +131,12 @@ public static bool TryGetSourceRepository(string remote, string branch, out stri
                     sourceBranch = contributionBranch;
                 }
 
-                if (sourceBranch != "live")
+                if (sourceBranch != "master" &&
+                    !GitUtility.RemoteBranchExists(sourceRemote, sourceBranch))
```
Comment (yufeih) [inline src/docfx/localization/LocalizationUtility.cs]: `ls-remote` is a network call, better not to access network in a utility function, this function is used in build as well.

Comment (OsmondJiang) [reply src/docfx/localization/LocalizationUtility.cs]: docfx make sure the source repo are only get once during restore and build. so it's should be OK

**Comment by OsmondJiang** [comment]
this PR current is waiting for #4386

**Comment by OsmondJiang** [comment]
> this PR current is waiting for #4386

Done

File: src/docfx/lib/git/GitUtility.cs [hunk 2]
```diff
@@ -129,6 +129,26 @@ public static string[] GetCommits(string path, string committish, int top = 1)
...
+            try
+            {
+                if (GitRemoteProxy != null)
+                {
+                    remote = GitRemoteProxy(remote);
+                }
+
+                return Execute(".", $"ls-remote --heads \"{remote}\" {branch}").Split('\n', StringSplitOptions.RemoveEmptyEntries).Any();
+            }
+            catch
```
Comment (yufeih) [inline src/docfx/lib/git/GitUtility.cs]: when is there an exception? can it be implemented without using exception handling

File: test/docfx.Test/build/ConfigTest.cs [hunk 3]
```diff
@@ -90,10 +90,6 @@ public static void LocConfigConventionTheme(string theme, string locale, string
         [InlineData("https://test.visualstudio.com/_git/abc.bs-Cyrl-BA", "master", "https://test.visualstudio.com/_git/abc", "master", "bs-cyrl-ba")]
         [InlineData("https://github.com/docs.zh-cn", "master-sxs", "https://github.com/docs", "master", "zh-cn")]
         [InlineData("https://github.com/docs.loc", "master-sxs.zh-cn", "https://github.com/docs", "master", "zh-cn")]
-        [InlineData("https://github.com/docs.zh-cn", "test", "https://github.com/docs", "master", "zh-cn")]
-        [InlineData("https://github.com/docs.loc", "test.zh-cn", "https://github.com/docs", "master", "zh-cn")]
-        [InlineData("https://github.com/docs.zh-cn", "test-sxs", "https://github.com/docs", "master", "zh-cn")]
-        [InlineData("https://github.com/docs.loc", "test-sxs.zh-cn", "https://github.com/docs", "master", "zh-cn")]
         public static void LocConfigConventionSourceRepo(string remote, string branch, string expectedSourceRemote, string expectedSourceBranch, string expectedLocale)
```
Comment (yufeih) [inline test/docfx.Test/build/ConfigTest.cs]: why remove these tests?

Comment (OsmondJiang) [reply test/docfx.Test/build/ConfigTest.cs]: fallback logic is not covered by TryGetSourceRepo method, it's controlled by restore/build pipeline using different methods

File: src/docfx/lib/git/GitUtility.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/lib/git/GitUtility.cs]: when the remote doesn't exists, just return false, I think `the remote doesn't exist` error should be thrown by `restore`

File: test/docfx.Test/build/ConfigTest.cs (hunk 3 above)
Comment (OsmondJiang) [reply test/docfx.Test/build/ConfigTest.cs]: we have spec test to cover that case

**Pull request merged**